// @dhis2/analytics/build/visualizations/config/adapters/dhis_highcharts/yAxis/gauge.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _arrayClean = _interopRequireDefault(require("d2-utilizr/lib/arrayClean"));

var _arraySort = _interopRequireDefault(require("d2-utilizr/lib/arraySort"));

var _isNumber = _interopRequireDefault(require("d2-utilizr/lib/isNumber"));

var _isObject = _interopRequireDefault(require("d2-utilizr/lib/isObject"));

var _objectClean = _interopRequireDefault(require("d2-utilizr/lib/objectClean"));

var _d2I18n = _interopRequireDefault(require("@dhis2/d2-i18n"));

var DEFAULT_MAX_VALUE = 100;
var DEFAULT_PLOT_LINE_STYLE = {
  zIndex: 5,
  width: 1,
  color: '#000'
};

var DEFAULT_TARGET_LINE_LABEL = _d2I18n.default.t('Target');

var DEFAULT_BASE_LINE_LABEL = _d2I18n.default.t('Base');

function getStopsByLegendSet(legendSet) {
  return (0, _isObject.default)(legendSet) ? (0, _arraySort.default)(legendSet.legends, 'ASC', 'endValue').map(function (legend) {
    return [parseFloat(legend.endValue) / DEFAULT_MAX_VALUE, legend.color];
  }) : undefined;
}

function getPlotLine(value, label) {
  return (0, _objectSpread2.default)({
    value: value
  }, DEFAULT_PLOT_LINE_STYLE, label && {
    label: {
      text: "".concat(label, ": ").concat(value)
    }
  });
}

function _default(layout, series, legendSet) {
  var plotLines = (0, _arrayClean.default)([(0, _isNumber.default)(layout.baseLineValue) ? getPlotLine(layout.baseLineValue, layout.baseLineLabel || DEFAULT_BASE_LINE_LABEL) : null, (0, _isNumber.default)(layout.targetLineValue) ? getPlotLine(layout.targetLineValue, layout.targetLineLabel || DEFAULT_TARGET_LINE_LABEL) : null]);
  return (0, _objectClean.default)((0, _objectSpread2.default)({
    min: (0, _isNumber.default)(layout.rangeAxisMinValue) ? layout.rangeAxisMinValue : 0,
    max: (0, _isNumber.default)(layout.rangeAxisMaxValue) ? layout.rangeAxisMaxValue : DEFAULT_MAX_VALUE,
    lineWidth: 0,
    minorTickInterval: null,
    tickLength: 0,
    tickAmount: 0,
    labels: {
      y: 18,
      style: {
        fontSize: '13px'
      }
    },
    title: {
      text: series[0].name
    },
    stops: getStopsByLegendSet(legendSet)
  }, plotLines.length && {
    plotLines: plotLines
  }));
}
