// @dhis2/analytics/build/modules/pivotTable/useTableClipping.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTableClipping = void 0;

var _react = require("react");

var _useScrollPosition = require("./useScrollPosition");

var _clipAxis = require("./clipAxis");

var useTableClipping = function useTableClipping(_ref) {
  var containerRef = _ref.containerRef,
      width = _ref.width,
      height = _ref.height,
      engine = _ref.engine,
      visualization = _ref.visualization;
  var scrollPosition = (0, _useScrollPosition.useScrollPosition)(containerRef);
  var rows = (0, _react.useMemo)(function () {
    return (0, _clipAxis.clipAxis)({
      position: scrollPosition.y,
      size: height,
      step: 25,
      totalCount: engine.height,
      headerCount: visualization.columns.length
    });
  }, [height, engine.height, scrollPosition.y, visualization.columns.length]);
  var columns = (0, _react.useMemo)(function () {
    return (0, _clipAxis.clipAxis)({
      position: scrollPosition.x,
      size: width,
      step: 150,
      totalCount: engine.width,
      headerCount: visualization.rows.length
    });
  }, [width, engine.width, scrollPosition.x, visualization.rows.length]);
  return {
    rows: rows,
    columns: columns
  };
};

exports.useTableClipping = useTableClipping;
