// @dhis2/analytics/build/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DataDimension", {
  enumerable: true,
  get: function get() {
    return _DataDimension.default;
  }
});
Object.defineProperty(exports, "PeriodDimension", {
  enumerable: true,
  get: function get() {
    return _PeriodDimension.default;
  }
});
Object.defineProperty(exports, "OrgUnitDimension", {
  enumerable: true,
  get: function get() {
    return _OrgUnitDimension.default;
  }
});
Object.defineProperty(exports, "DynamicDimension", {
  enumerable: true,
  get: function get() {
    return _DynamicDimension.default;
  }
});
Object.defineProperty(exports, "DimensionsPanel", {
  enumerable: true,
  get: function get() {
    return _DimensionsPanel.default;
  }
});
Object.defineProperty(exports, "DimensionItem", {
  enumerable: true,
  get: function get() {
    return _DimensionItem.default;
  }
});
Object.defineProperty(exports, "DimensionFilter", {
  enumerable: true,
  get: function get() {
    return _Filter.default;
  }
});
Object.defineProperty(exports, "ItemSelector", {
  enumerable: true,
  get: function get() {
    return _ItemSelector.default;
  }
});
Object.defineProperty(exports, "DimensionMenu", {
  enumerable: true,
  get: function get() {
    return _DimensionMenu.default;
  }
});
Object.defineProperty(exports, "PivotTable", {
  enumerable: true,
  get: function get() {
    return _PivotTable.default;
  }
});
Object.defineProperty(exports, "apiFetchDimensions", {
  enumerable: true,
  get: function get() {
    return _dimensions.apiFetchDimensions;
  }
});
Object.defineProperty(exports, "apiFetchRecommendedIds", {
  enumerable: true,
  get: function get() {
    return _dimensions.apiFetchRecommendedIds;
  }
});
Object.defineProperty(exports, "getAxisName", {
  enumerable: true,
  get: function get() {
    return _axis.getAxisName;
  }
});
Object.defineProperty(exports, "getAxisNameByLayoutType", {
  enumerable: true,
  get: function get() {
    return _axis.getAxisNameByLayoutType;
  }
});
Object.defineProperty(exports, "DIMENSION_ID_DATA", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.DIMENSION_ID_DATA;
  }
});
Object.defineProperty(exports, "DIMENSION_ID_PERIOD", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.DIMENSION_ID_PERIOD;
  }
});
Object.defineProperty(exports, "DIMENSION_ID_ORGUNIT", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.DIMENSION_ID_ORGUNIT;
  }
});
Object.defineProperty(exports, "DIMENSION_ID_ASSIGNED_CATEGORIES", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.DIMENSION_ID_ASSIGNED_CATEGORIES;
  }
});
Object.defineProperty(exports, "DIMENSION_PROP_NO_ITEMS", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.DIMENSION_PROP_NO_ITEMS;
  }
});
Object.defineProperty(exports, "filterOutPredefinedDimensions", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.filterOutPredefinedDimensions;
  }
});
Object.defineProperty(exports, "getPredefinedDimensionProp", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.getPredefinedDimensionProp;
  }
});
Object.defineProperty(exports, "getDimensionById", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.getDimensionById;
  }
});
Object.defineProperty(exports, "getPredefinedDimensions", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.getPredefinedDimensions;
  }
});
Object.defineProperty(exports, "getFixedDimensions", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.getFixedDimensions;
  }
});
Object.defineProperty(exports, "getDynamicDimensions", {
  enumerable: true,
  get: function get() {
    return _predefinedDimensions.getDynamicDimensions;
  }
});
Object.defineProperty(exports, "ouIdHelper", {
  enumerable: true,
  get: function get() {
    return _ouIdHelper.ouIdHelper;
  }
});
Object.defineProperty(exports, "LAYOUT", {
  enumerable: true,
  get: function get() {
    return _layout.LAYOUT;
  }
});
Object.defineProperty(exports, "layoutFilterDimensions", {
  enumerable: true,
  get: function get() {
    return _layoutFilterDimensions.layoutFilterDimensions;
  }
});
Object.defineProperty(exports, "layoutGetAllAxes", {
  enumerable: true,
  get: function get() {
    return _layoutGetAllAxes.layoutGetAllAxes;
  }
});
Object.defineProperty(exports, "layoutGetAllDimensions", {
  enumerable: true,
  get: function get() {
    return _layoutGetAllDimensions.layoutGetAllDimensions;
  }
});
Object.defineProperty(exports, "layoutGetAllItemIds", {
  enumerable: true,
  get: function get() {
    return _layoutGetAllItemIds.layoutGetAllItemIds;
  }
});
Object.defineProperty(exports, "layoutGetAllItems", {
  enumerable: true,
  get: function get() {
    return _layoutGetAllItems.layoutGetAllItems;
  }
});
Object.defineProperty(exports, "layoutGetAxisIdDimensionIdsObject", {
  enumerable: true,
  get: function get() {
    return _layoutGetAxisIdDimensionIdsObject.layoutGetAxisIdDimensionIdsObject;
  }
});
Object.defineProperty(exports, "layoutGetDimension", {
  enumerable: true,
  get: function get() {
    return _layoutGetDimension.layoutGetDimension;
  }
});
Object.defineProperty(exports, "layoutGetDimensionItems", {
  enumerable: true,
  get: function get() {
    return _layoutGetDimensionItems.layoutGetDimensionItems;
  }
});
Object.defineProperty(exports, "layoutReplaceDimension", {
  enumerable: true,
  get: function get() {
    return _layoutReplaceDimension.layoutReplaceDimension;
  }
});
Object.defineProperty(exports, "layoutGetDimensionIdItemIdsObject", {
  enumerable: true,
  get: function get() {
    return _layoutGetDimensionIdItemIdsObject.layoutGetDimensionIdItemIdsObject;
  }
});
Object.defineProperty(exports, "layoutHasDataDimension", {
  enumerable: true,
  get: function get() {
    return _layoutHasDataDimension.layoutHasDataDimension;
  }
});
Object.defineProperty(exports, "layoutHasDimension", {
  enumerable: true,
  get: function get() {
    return _layoutHasDimension.layoutHasDimension;
  }
});
Object.defineProperty(exports, "layoutHasDynamicDimension", {
  enumerable: true,
  get: function get() {
    return _layoutHasDynamicDimension.layoutHasDynamicDimension;
  }
});
Object.defineProperty(exports, "layoutHasPeriodDimension", {
  enumerable: true,
  get: function get() {
    return _layoutHasPeriodDimension.layoutHasPeriodDimension;
  }
});
Object.defineProperty(exports, "AXIS", {
  enumerable: true,
  get: function get() {
    return _axis2.AXIS;
  }
});
Object.defineProperty(exports, "AXIS_ID_COLUMNS", {
  enumerable: true,
  get: function get() {
    return _axis2.AXIS_ID_COLUMNS;
  }
});
Object.defineProperty(exports, "AXIS_ID_ROWS", {
  enumerable: true,
  get: function get() {
    return _axis2.AXIS_ID_ROWS;
  }
});
Object.defineProperty(exports, "AXIS_ID_FILTERS", {
  enumerable: true,
  get: function get() {
    return _axis2.AXIS_ID_FILTERS;
  }
});
Object.defineProperty(exports, "DEFAULT_AXIS_IDS", {
  enumerable: true,
  get: function get() {
    return _axis2.DEFAULT_AXIS_IDS;
  }
});
Object.defineProperty(exports, "axisGetAllItems", {
  enumerable: true,
  get: function get() {
    return _axisGetAllItems.axisGetAllItems;
  }
});
Object.defineProperty(exports, "axisGetDimension", {
  enumerable: true,
  get: function get() {
    return _axisGetDimension.axisGetDimension;
  }
});
Object.defineProperty(exports, "axisGetDimensionIds", {
  enumerable: true,
  get: function get() {
    return _axisGetDimensionIds.axisGetDimensionIds;
  }
});
Object.defineProperty(exports, "axisHasDataDimension", {
  enumerable: true,
  get: function get() {
    return _axisHasDataDimension.axisHasDataDimension;
  }
});
Object.defineProperty(exports, "axisHasDimension", {
  enumerable: true,
  get: function get() {
    return _axisHasDimension.axisHasDimension;
  }
});
Object.defineProperty(exports, "axisHasPeriodDimension", {
  enumerable: true,
  get: function get() {
    return _axisHasPeriodDimension.axisHasPeriodDimension;
  }
});
Object.defineProperty(exports, "axisHasOuDimension", {
  enumerable: true,
  get: function get() {
    return _axisHasOuDimension.axisHasOuDimension;
  }
});
Object.defineProperty(exports, "axisIsEmpty", {
  enumerable: true,
  get: function get() {
    return _axisIsEmpty.axisIsEmpty;
  }
});
Object.defineProperty(exports, "DIMENSION", {
  enumerable: true,
  get: function get() {
    return _dimension.DIMENSION;
  }
});
Object.defineProperty(exports, "DIMENSION_PROP_ID", {
  enumerable: true,
  get: function get() {
    return _dimension.DIMENSION_PROP_ID;
  }
});
Object.defineProperty(exports, "DIMENSION_PROP_ITEMS", {
  enumerable: true,
  get: function get() {
    return _dimension.DIMENSION_PROP_ITEMS;
  }
});
Object.defineProperty(exports, "DIMENSION_PROPS", {
  enumerable: true,
  get: function get() {
    return _dimension.DIMENSION_PROPS;
  }
});
Object.defineProperty(exports, "dimensionCreate", {
  enumerable: true,
  get: function get() {
    return _dimensionCreate.dimensionCreate;
  }
});
Object.defineProperty(exports, "dimensionGetId", {
  enumerable: true,
  get: function get() {
    return _dimensionGetId.dimensionGetId;
  }
});
Object.defineProperty(exports, "dimensionGetItemIds", {
  enumerable: true,
  get: function get() {
    return _dimensionGetItemIds.dimensionGetItemIds;
  }
});
Object.defineProperty(exports, "dimensionGetItems", {
  enumerable: true,
  get: function get() {
    return _dimensionGetItems.dimensionGetItems;
  }
});
Object.defineProperty(exports, "dimensionIs", {
  enumerable: true,
  get: function get() {
    return _dimensionIs.dimensionIs;
  }
});
Object.defineProperty(exports, "dimensionIsEmpty", {
  enumerable: true,
  get: function get() {
    return _dimensionIsEmpty.dimensionIsEmpty;
  }
});
Object.defineProperty(exports, "dimensionIsValid", {
  enumerable: true,
  get: function get() {
    return _dimensionIsValid.dimensionIsValid;
  }
});
Object.defineProperty(exports, "ITEM", {
  enumerable: true,
  get: function get() {
    return _item.ITEM;
  }
});
Object.defineProperty(exports, "ITEM_PROP_ID", {
  enumerable: true,
  get: function get() {
    return _item.ITEM_PROP_ID;
  }
});
Object.defineProperty(exports, "ITEM_PROPS", {
  enumerable: true,
  get: function get() {
    return _item.ITEM_PROPS;
  }
});
Object.defineProperty(exports, "itemGetId", {
  enumerable: true,
  get: function get() {
    return _itemGetId.itemGetId;
  }
});
Object.defineProperty(exports, "itemIsValid", {
  enumerable: true,
  get: function get() {
    return _itemIsValid.itemIsValid;
  }
});
Object.defineProperty(exports, "getLayoutTypeByVisType", {
  enumerable: true,
  get: function get() {
    return _visTypeToLayoutType.getLayoutTypeByVisType;
  }
});
Object.defineProperty(exports, "VIS_TYPE_COLUMN", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_COLUMN;
  }
});
Object.defineProperty(exports, "VIS_TYPE_STACKED_COLUMN", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_STACKED_COLUMN;
  }
});
Object.defineProperty(exports, "VIS_TYPE_BAR", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_BAR;
  }
});
Object.defineProperty(exports, "VIS_TYPE_STACKED_BAR", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_STACKED_BAR;
  }
});
Object.defineProperty(exports, "VIS_TYPE_LINE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_LINE;
  }
});
Object.defineProperty(exports, "VIS_TYPE_AREA", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_AREA;
  }
});
Object.defineProperty(exports, "VIS_TYPE_PIE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_PIE;
  }
});
Object.defineProperty(exports, "VIS_TYPE_RADAR", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_RADAR;
  }
});
Object.defineProperty(exports, "VIS_TYPE_GAUGE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_GAUGE;
  }
});
Object.defineProperty(exports, "VIS_TYPE_BUBBLE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_BUBBLE;
  }
});
Object.defineProperty(exports, "VIS_TYPE_YEAR_OVER_YEAR_LINE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_YEAR_OVER_YEAR_LINE;
  }
});
Object.defineProperty(exports, "VIS_TYPE_YEAR_OVER_YEAR_COLUMN", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_YEAR_OVER_YEAR_COLUMN;
  }
});
Object.defineProperty(exports, "VIS_TYPE_SINGLE_VALUE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_SINGLE_VALUE;
  }
});
Object.defineProperty(exports, "VIS_TYPE_PIVOT_TABLE", {
  enumerable: true,
  get: function get() {
    return _visTypes.VIS_TYPE_PIVOT_TABLE;
  }
});
Object.defineProperty(exports, "visTypeDisplayNames", {
  enumerable: true,
  get: function get() {
    return _visTypes.visTypeDisplayNames;
  }
});
Object.defineProperty(exports, "visTypeIcons", {
  enumerable: true,
  get: function get() {
    return _visTypes.visTypeIcons;
  }
});
Object.defineProperty(exports, "getDisplayNameByVisType", {
  enumerable: true,
  get: function get() {
    return _visTypes.getDisplayNameByVisType;
  }
});
Object.defineProperty(exports, "defaultChartType", {
  enumerable: true,
  get: function get() {
    return _visTypes.defaultChartType;
  }
});
Object.defineProperty(exports, "isStacked", {
  enumerable: true,
  get: function get() {
    return _visTypes.isStacked;
  }
});
Object.defineProperty(exports, "isYearOverYear", {
  enumerable: true,
  get: function get() {
    return _visTypes.isYearOverYear;
  }
});
Object.defineProperty(exports, "isDualAxisType", {
  enumerable: true,
  get: function get() {
    return _visTypes.isDualAxisType;
  }
});
Object.defineProperty(exports, "isSingleValue", {
  enumerable: true,
  get: function get() {
    return _visTypes.isSingleValue;
  }
});
Object.defineProperty(exports, "LAYOUT_TYPE_DEFAULT", {
  enumerable: true,
  get: function get() {
    return _layoutTypes.LAYOUT_TYPE_DEFAULT;
  }
});
Object.defineProperty(exports, "LAYOUT_TYPE_PIE", {
  enumerable: true,
  get: function get() {
    return _layoutTypes.LAYOUT_TYPE_PIE;
  }
});
Object.defineProperty(exports, "LAYOUT_TYPE_SINGLE_VALUE", {
  enumerable: true,
  get: function get() {
    return _layoutTypes.LAYOUT_TYPE_SINGLE_VALUE;
  }
});
Object.defineProperty(exports, "LAYOUT_TYPE_YEAR_OVER_YEAR", {
  enumerable: true,
  get: function get() {
    return _layoutTypes.LAYOUT_TYPE_YEAR_OVER_YEAR;
  }
});
Object.defineProperty(exports, "LAYOUT_TYPE_PIVOT_TABLE", {
  enumerable: true,
  get: function get() {
    return _layoutTypes.LAYOUT_TYPE_PIVOT_TABLE;
  }
});
Object.defineProperty(exports, "getAvailableAxes", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getAvailableAxes;
  }
});
Object.defineProperty(exports, "getDisallowedDimensions", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getDisallowedDimensions;
  }
});
Object.defineProperty(exports, "getAxisMaxNumberOfItems", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getAxisMaxNumberOfItems;
  }
});
Object.defineProperty(exports, "getAxisMaxNumberOfDimensions", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getAxisMaxNumberOfDimensions;
  }
});
Object.defineProperty(exports, "getAxisMinNumberOfDimensions", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getAxisMinNumberOfDimensions;
  }
});
Object.defineProperty(exports, "hasAxisTooManyItems", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.hasAxisTooManyItems;
  }
});
Object.defineProperty(exports, "getAxisPerLockedDimension", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getAxisPerLockedDimension;
  }
});
Object.defineProperty(exports, "getAllLockedDimensionIds", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getAllLockedDimensionIds;
  }
});
Object.defineProperty(exports, "canDimensionBeAddedToAxis", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.canDimensionBeAddedToAxis;
  }
});
Object.defineProperty(exports, "isDimensionLocked", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.isDimensionLocked;
  }
});
Object.defineProperty(exports, "isAxisFull", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.isAxisFull;
  }
});
Object.defineProperty(exports, "getTransferableDimension", {
  enumerable: true,
  get: function get() {
    return _layoutUiRules.getTransferableDimension;
  }
});
Object.defineProperty(exports, "createVisualization", {
  enumerable: true,
  get: function get() {
    return _visualizations.createVisualization;
  }
});

var _DataDimension = _interopRequireDefault(require("./components/DataDimension/DataDimension"));

var _PeriodDimension = _interopRequireDefault(require("./components/PeriodDimension/PeriodDimension"));

var _OrgUnitDimension = _interopRequireDefault(require("./components/OrgUnitDimension/OrgUnitDimension"));

var _DynamicDimension = _interopRequireDefault(require("./components/DynamicDimension/DynamicDimension"));

var _DimensionsPanel = _interopRequireDefault(require("./components/DimensionsPanel/DimensionsPanel"));

var _DimensionItem = _interopRequireDefault(require("./components/DimensionsPanel/List/DimensionItem"));

var _Filter = _interopRequireDefault(require("./components/Filter/Filter"));

var _ItemSelector = _interopRequireDefault(require("./components/ItemSelector/ItemSelector"));

var _DimensionMenu = _interopRequireDefault(require("./components/DimensionMenu"));

var _PivotTable = _interopRequireDefault(require("./components/PivotTable/PivotTable"));

var _dimensions = require("./api/dimensions");

var _axis = require("./modules/axis");

var _predefinedDimensions = require("./modules/predefinedDimensions");

var _ouIdHelper = require("./modules/ouIdHelper");

var _layout = require("./modules/layout/layout");

var _layoutFilterDimensions = require("./modules/layout/layoutFilterDimensions");

var _layoutGetAllAxes = require("./modules/layout/layoutGetAllAxes");

var _layoutGetAllDimensions = require("./modules/layout/layoutGetAllDimensions");

var _layoutGetAllItemIds = require("./modules/layout/layoutGetAllItemIds");

var _layoutGetAllItems = require("./modules/layout/layoutGetAllItems");

var _layoutGetAxisIdDimensionIdsObject = require("./modules/layout/layoutGetAxisIdDimensionIdsObject");

var _layoutGetDimension = require("./modules/layout/layoutGetDimension");

var _layoutGetDimensionItems = require("./modules/layout/layoutGetDimensionItems");

var _layoutReplaceDimension = require("./modules/layout/layoutReplaceDimension");

var _layoutGetDimensionIdItemIdsObject = require("./modules/layout/layoutGetDimensionIdItemIdsObject");

var _layoutHasDataDimension = require("./modules/layout/layoutHasDataDimension");

var _layoutHasDimension = require("./modules/layout/layoutHasDimension");

var _layoutHasDynamicDimension = require("./modules/layout/layoutHasDynamicDimension");

var _layoutHasPeriodDimension = require("./modules/layout/layoutHasPeriodDimension");

var _axis2 = require("./modules/layout/axis");

var _axisGetAllItems = require("./modules/layout/axisGetAllItems");

var _axisGetDimension = require("./modules/layout/axisGetDimension");

var _axisGetDimensionIds = require("./modules/layout/axisGetDimensionIds");

var _axisHasDataDimension = require("./modules/layout/axisHasDataDimension");

var _axisHasDimension = require("./modules/layout/axisHasDimension");

var _axisHasPeriodDimension = require("./modules/layout/axisHasPeriodDimension");

var _axisHasOuDimension = require("./modules/layout/axisHasOuDimension");

var _axisIsEmpty = require("./modules/layout/axisIsEmpty");

var _dimension = require("./modules/layout/dimension");

var _dimensionCreate = require("./modules/layout/dimensionCreate");

var _dimensionGetId = require("./modules/layout/dimensionGetId");

var _dimensionGetItemIds = require("./modules/layout/dimensionGetItemIds");

var _dimensionGetItems = require("./modules/layout/dimensionGetItems");

var _dimensionIs = require("./modules/layout/dimensionIs");

var _dimensionIsEmpty = require("./modules/layout/dimensionIsEmpty");

var _dimensionIsValid = require("./modules/layout/dimensionIsValid");

var _item = require("./modules/layout/item");

var _itemGetId = require("./modules/layout/itemGetId");

var _itemIsValid = require("./modules/layout/itemIsValid");

var _visTypeToLayoutType = require("./modules/visTypeToLayoutType");

var _visTypes = require("./modules/visTypes");

var _layoutTypes = require("./modules/layoutTypes");

var _layoutUiRules = require("./modules/layoutUiRules");

var _visualizations = require("./visualizations");
