// @dhis2/analytics/build/components/PivotTable/PivotTableValueCell.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PivotTableValueCell = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PivotTable = require("./styles/PivotTable.style");

var PivotTableValueCell = function PivotTableValueCell(_ref) {
  var engine = _ref.engine,
      row = _ref.row,
      column = _ref.column;
  var value = engine.get({
    row: row,
    column: column,
    field: 'value'
  });
  var type = engine.getCellType({
    row: row,
    column: column
  });
  return _react.default.createElement("td", {
    key: column,
    className: "jsx-".concat(_PivotTable.cell.__hash) + " " + (type || "")
  }, _react.default.createElement(_style.default, {
    id: _PivotTable.cell.__hash
  }, _PivotTable.cell), value || null);
};

exports.PivotTableValueCell = PivotTableValueCell;
PivotTableValueCell.propTypes = {
  column: _propTypes.default.number.isRequired,
  engine: _propTypes.default.object.isRequired,
  row: _propTypes.default.number.isRequired
};
