// @dhis2/analytics/build/components/PivotTable/PivotTableColumnHeaders.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PivotTableColumnHeaders = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PivotTableClippedAxis = require("./PivotTableClippedAxis");

var _PivotTableColumnHeaderCell = require("./PivotTableColumnHeaderCell");

var _PivotTableDimensionLabelCell = require("./PivotTableDimensionLabelCell");

var PivotTableColumnHeaders = function PivotTableColumnHeaders(_ref) {
  var engine = _ref.engine,
      clippingResult = _ref.clippingResult;

  var _ref2 =
  /*#__PURE__*/
  _react.default.createElement(_PivotTableDimensionLabelCell.PivotTableDimensionLabelCell, {
    engine: engine
  });

  return engine.dimensionLookup.columns.map(function (_, level) {
    return _react.default.createElement("tr", {
      key: level
    }, _ref2, _react.default.createElement(_PivotTableClippedAxis.PivotTableClippedAxis, {
      axisClippingResult: clippingResult.columns,
      EmptyComponent: function EmptyComponent(_ref3) {
        var size = _ref3.size;
        return _react.default.createElement("th", {
          style: {
            minWidth: size
          }
        });
      },
      ItemComponent: function ItemComponent(_ref4) {
        var index = _ref4.index;
        return _react.default.createElement(_PivotTableColumnHeaderCell.PivotTableColumnHeaderCell, {
          engine: engine,
          clippingResult: clippingResult,
          index: index,
          level: level
        });
      }
    }));
  });
};

exports.PivotTableColumnHeaders = PivotTableColumnHeaders;
PivotTableColumnHeaders.propTypes = {
  clippingResult: _propTypes.default.shape({
    columns: _propTypes.default.object.isRequired
  }).isRequired,
  engine: _propTypes.default.object.isRequired
};
