// @dhis2/analytics/build/components/PivotTable/PivotTableColumnHeaderCell.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PivotTableColumnHeaderCell = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PivotTableHeaderCell = require("./PivotTableHeaderCell");

var _PivotTable = require("./styles/PivotTable.style");

var PivotTableColumnHeaderCell = function PivotTableColumnHeaderCell(_ref) {
  var engine = _ref.engine,
      clippingResult = _ref.clippingResult,
      index = _ref.index,
      level = _ref.level;
  return _react.default.createElement(_PivotTableHeaderCell.PivotTableHeaderCell, {
    axisClippingResult: clippingResult.columns,
    index: index,
    level: level,
    getHeader: function getHeader(idx) {
      return engine.getColumnHeader(idx);
    },
    render: function render(header) {
      return _react.default.createElement("th", {
        colSpan: header.span,
        className: "jsx-".concat(_PivotTable.cell.__hash) + " " + ((header.name && header.name !== 'TOTAL' ? 'col-header' : 'empty-header') || "")
      }, _react.default.createElement(_style.default, {
        id: _PivotTable.cell.__hash
      }, _PivotTable.cell), header.name);
    }
  });
};

exports.PivotTableColumnHeaderCell = PivotTableColumnHeaderCell;
PivotTableColumnHeaderCell.propTypes = {
  clippingResult: _propTypes.default.shape({
    columns: _propTypes.default.object.isRequired
  }).isRequired,
  engine: _propTypes.default.object.isRequired,
  index: _propTypes.default.number.isRequired,
  level: _propTypes.default.number.isRequired
};
