// @dhis2/analytics/build/components/PivotTable/PivotTable.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PivotTable = require("./styles/PivotTable.style");

var _PivotTableEngine = require("../../modules/pivotTable/PivotTableEngine");

var _useParentSize2 = require("../../modules/pivotTable/useParentSize");

var _PivotTableColumnHeaders = require("./PivotTableColumnHeaders");

var _useTableClipping = require("../../modules/pivotTable/useTableClipping");

var _PivotTableRow = require("./PivotTableRow");

var _PivotTableClippedAxis = require("./PivotTableClippedAxis");

var PivotTable = function PivotTable(_ref) {
  var visualization = _ref.visualization,
      data = _ref.data;
  var containerRef = (0, _react.useRef)(undefined);

  var _useParentSize = (0, _useParentSize2.useParentSize)(containerRef),
      width = _useParentSize.width,
      height = _useParentSize.height;

  var engine = (0, _react.useMemo)(function () {
    return new _PivotTableEngine.PivotTableEngine(visualization, data);
  }, [visualization, data]);
  var clippingResult = (0, _useTableClipping.useTableClipping)({
    containerRef: containerRef,
    width: width,
    height: height,
    engine: engine,
    visualization: visualization
  });
  return _react.default.createElement("div", {
    style: {
      width: width,
      height: height
    },
    ref: containerRef,
    className: "jsx-".concat(_PivotTable.table.__hash) + " " + "pivot-table-container"
  }, _react.default.createElement(_style.default, {
    id: _PivotTable.table.__hash
  }, _PivotTable.table), width === 0 || height === 0 ? null : _react.default.createElement("table", {
    className: "jsx-".concat(_PivotTable.table.__hash)
  }, _react.default.createElement("thead", {
    className: "jsx-".concat(_PivotTable.table.__hash)
  }, _react.default.createElement(_PivotTableColumnHeaders.PivotTableColumnHeaders, {
    engine: engine,
    clippingResult: clippingResult
  })), _react.default.createElement("tbody", {
    className: "jsx-".concat(_PivotTable.table.__hash)
  }, _react.default.createElement(_PivotTableClippedAxis.PivotTableClippedAxis, {
    axisClippingResult: clippingResult.rows,
    EmptyComponent: function EmptyComponent(_ref2) {
      var size = _ref2.size;
      return _react.default.createElement("tr", {
        className: "jsx-".concat(_PivotTable.table.__hash)
      }, _react.default.createElement("td", {
        style: {
          height: size
        },
        className: "jsx-".concat(_PivotTable.table.__hash)
      }));
    },
    ItemComponent: function ItemComponent(_ref3) {
      var index = _ref3.index;
      return _react.default.createElement(_PivotTableRow.PivotTableRow, {
        key: index,
        engine: engine,
        clippingResult: clippingResult,
        rowIndex: index
      });
    }
  }))));
};

PivotTable.propTypes = {
  data: _propTypes.default.object.isRequired,
  visualization: _propTypes.default.object.isRequired
};
var _default = PivotTable;
exports.default = _default;
