// @dhis2/analytics/build/components/PivotTable/PivotTableRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PivotTableRow = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PivotTableClippedAxis = require("./PivotTableClippedAxis");

var _PivotTableRowHeaderCell = require("./PivotTableRowHeaderCell");

var _PivotTableValueCell = require("./PivotTableValueCell");

var _PivotTableEmptyCell = require("./PivotTableEmptyCell");

var _ref2 =
/*#__PURE__*/
_react.default.createElement(_PivotTableEmptyCell.PivotTableEmptyCell, {
  type: "value"
});

var PivotTableRow = function PivotTableRow(_ref) {
  var engine = _ref.engine,
      clippingResult = _ref.clippingResult,
      rowIndex = _ref.rowIndex;
  return _react.default.createElement("tr", null, engine.dimensionLookup.rows.map(function (_, rowLevel) {
    return _react.default.createElement(_PivotTableRowHeaderCell.PivotTableRowHeaderCell, {
      key: rowLevel,
      engine: engine,
      clippingResult: clippingResult,
      rowIndex: rowIndex,
      rowLevel: rowLevel
    });
  }), _react.default.createElement(_PivotTableClippedAxis.PivotTableClippedAxis, {
    axisClippingResult: clippingResult.columns,
    EmptyComponent: function EmptyComponent() {
      return _ref2;
    },
    ItemComponent: function ItemComponent(_ref3) {
      var columnIndex = _ref3.index;
      return _react.default.createElement(_PivotTableValueCell.PivotTableValueCell, {
        engine: engine,
        row: rowIndex,
        column: columnIndex
      });
    }
  }));
};

exports.PivotTableRow = PivotTableRow;
PivotTableRow.propTypes = {
  clippingResult: _propTypes.default.shape({
    columns: _propTypes.default.object.isRequired,
    rows: _propTypes.default.object.isRequired
  }).isRequired,
  engine: _propTypes.default.object.isRequired,
  rowIndex: _propTypes.default.number.isRequired
};
