// @dhis2/analytics/build/components/PivotTable/PivotTableRowHeaderCell.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PivotTableRowHeaderCell = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PivotTableHeaderCell = require("./PivotTableHeaderCell");

var _PivotTable = require("./styles/PivotTable.style");

var PivotTableRowHeaderCell = function PivotTableRowHeaderCell(_ref) {
  var engine = _ref.engine,
      clippingResult = _ref.clippingResult,
      rowIndex = _ref.rowIndex,
      rowLevel = _ref.rowLevel;
  return _react.default.createElement(_PivotTableHeaderCell.PivotTableHeaderCell, {
    axisClippingResult: clippingResult.rows,
    index: rowIndex,
    level: rowLevel,
    getHeader: function getHeader(idx) {
      return engine.getRowHeader(idx);
    },
    render: function render(header) {
      return _react.default.createElement("td", {
        rowSpan: header.span,
        className: "jsx-".concat(_PivotTable.cell.__hash) + " " + ((header.name && header.name !== 'TOTAL' ? 'row-header' : 'empty-header') || "")
      }, _react.default.createElement(_style.default, {
        id: _PivotTable.cell.__hash
      }, _PivotTable.cell), header.name);
    }
  });
};

exports.PivotTableRowHeaderCell = PivotTableRowHeaderCell;
PivotTableRowHeaderCell.propTypes = {
  clippingResult: _propTypes.default.shape({
    rows: _propTypes.default.object.isRequired
  }).isRequired,
  engine: _propTypes.default.object.isRequired,
  rowIndex: _propTypes.default.number.isRequired,
  rowLevel: _propTypes.default.number.isRequired
};
